docs/source/checkpoints/dist_ckpt.rst

```diff
@@ -0,0 +1,392 @@
+This is the reason for introducing ShardedTensorFactories - we have to register the original model parameter as `ShardedTensorFactories.data` and apply any subsequent transformations as a factory function in order to make sure that the same transformation can be applied to the optimizer states.
+
+
+Tensors transformations
+=======================
+ShardedTensor API allows declaring some basic transformations that should be performed on the tensors during saving and loading.
+
+Shape mismatch
+--------------
+The `allow_shape_mismatch` flag allows to relax the requirement of matching global tensor shapes during loading.
+Extra padding is filled with zeros or stripped depending on the mismatch kind.
+This comes handy for layers like embedding which might be padded according to parallelism for performance reasons.
+
+Flattening
+----------
+The `flattened_range` attribute allows to declare the fact that `ShardedTensor.data` is actually a slice of a flattened model parameter.
+This corresponds to a transformation used in Distributed Optimizers which flattens the data and shards it along the data-parallel domain.
+
+Extra flattening comes with an efficiency challenge during checkpoint resharding.
+Since flattening is applied after the global tensors is sharded into the grid of local chunks, loading after resharding requires accessing incontiguous data fragments.
+An example solution for that is implemented in the `dist_checkpointing/strategies/resharding.py` module and involves saving the flattened tensor with a different global shape than the original one.
+
+Example: For a global tensor [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11]] with sharding by TP (tensor-parallel) over the second axis, here are the local shards if TP=2:
+
+.. list-table::
+   :widths: 50 50
+   :header-rows: 1
+
+   * - Rank
+     - Local shards
+   * - 0
+     - [[0, 1, 2], [6, 7, 8]]
+   * - 1
+     - [[3, 4, 5], [9, 10, 11]]
+
+After flattening and sharding by e.g. DP=3:
+
+.. list-table::
+   :widths: 50 50
+   :header-rows: 1
+
+   * - Rank
+     - Local shards
+   * - 0
+     - [0, 1]
+   * - 2
+     - [2, 6]
+   * - 4
+     - [7, 8]
+   * - 1
+     - [3, 4]
+   * - 3
+     - [5, 9]
+   * - 5
+     - [10, 11]
+
+The same tensor after sharding by TP=6, flattening and sharding by DP=1:
+
+
+.. list-table::
+   :widths: 50 50
+   :header-rows: 1
+
+   * - Rank
+     - Local shards
+   * - 0
+     - [0, 6]
+   * - 1
+     - [1, 7]
+   * - 2
+     - [2, 8]
+   * - 3
+     - [3, 9]
+   * - 4
+     - [4, 10]
+   * - 5
+     - [5, 11]
+
+
+Arbitrary transformations
+-------------------------
+The way to apply arbitrary transformations to the tensors during saving and loading is with ShardedTensorFactory, which allows to define such transformations as a function that can be reapplied to any ShardedTensor (in particular, a ShardedTensor representing optimizer states).
+Such "build" function is also tied to a "merge" function that can apply an inverse transformation during loading.
```

Comment:
To apply arbitrary transformations to tensors during saving and loading, the ShardedTensorFactory can be used. This factory allows the definition of transformations as functions that can be reapplied to any ShardedTensor, including those representing optimizer states. Specifically, the ‘build’ function defines the transformation, and it is paired with a corresponding ‘merge’ function that applies the inverse transformation during loading.